ngraph_bridge/ngraph_encapsulate_clusters.h

```diff
@@ -35,10 +39,107 @@ typedef std::map<std::string, std::vector<int>> ShapeHintMap;
+//                                       |
+//                                       v
+//                                 NewClusterIds
+// Any other order of calling will generate errors
+// Cannot be copied/moved or reset
+class Encapsulator {
+ public:
+  Encapsulator(Graph* g);
+  // Populate ClusterManager with the subgraphs for each potential encapsulate
+  Status AnalysisPass();
+  // Perform the actual graph surgery
+  Status RewritePass(
+      FunctionDefLibrary* fdeflib, int graph_id,
+      const std::unordered_map<std::string, std::string>& device_config);
+  Status NewClusterIds(std::set<int>& result);
+
+  Encapsulator(const Encapsulator&) = delete;
+  Encapsulator(Encapsulator&&) = delete;
+  Encapsulator& operator=(const Encapsulator&) = delete;
+  Encapsulator& operator=(Encapsulator&&) = delete;
+
+ private:
+  Graph* graph;
+  // boolean to indicate if analysis has been done
+  // If not rewritepass should not be called
+  bool analysis_done;
+  // boolean to indicate that rewrite is done;
+  bool rewrite_done;
+  // A map from cluster indices to the expected device name for nodes
+  // in that cluster.
+  std::map<int, std::string> device_name_map;
+
+  // We *should* eventually have a way of monitoring the device and the backend
+  // together
+  std::map<int, std::string> backend_name_map;
+
+  // As we build the graph we will be tracking the.. TODO(amprocte): finish
+  // this comment.
+  std::map<std::tuple<int, int>, std::tuple<int, int>> output_remap_map;
+  std::map<std::tuple<int, int, int>, int> input_remap_map;
+  std::map<std::tuple<int, std::string, int>, string> input_rename_map;
+
+  // A map from cluster indices to a vector of input data types.
+  std::map<int, std::vector<std::tuple<int, int, DataType>>> cluster_input_map;
+  // A map from cluster indices to a vector of output data types.
+  std::map<int, std::vector<DataType>> cluster_output_dt_map;
+
+  // A map from cluster indices to corresponding NGraphEncapsulate nodes.
+  std::map<int, Node*> cluster_node_map;
+
+  std::set<int> cluster_indices_for_this_graph;
+
+  static void AddInput(NodeDef* dst, StringPiece src_name, int src_slot);
+};
+
+// Translates TF subgraph to ng function then compiles it
+Status PerformAOTOnEncapsulates(Graph* graph, const AOTInfo& aot_info);
+
+std::string HintAsString(ShapeHintMap single_hint);
+
+// Given a node, partialshape info from TF (present in the .pb itself) and a
+// shape hint, combine all that information
+PartialShape CombineNodeInfoAndHint(Node* node,
+                                    PartialShape partial_shape_from_node,
+                                    const ShapeHintMap& single_hint);
+
+// Given a TF graph, it scans it for inputs and finds what TF is saying about
+// their shapes (in the .pb itself)
+// Creates a map between input node names and PartialShape information we get
+// from the TF graph
+std::map<std::string, PartialShape> GetShapesFromTFInputnodes(
```

Comment:
Similar comment for this. It should work only for a placeholder as it is invalid for any other node type? For eg, if i pass in  input_node_type as Add, the function will probably happily work. There is no check whether input_node_type is really a boundary/input node. 

GetShapesFromTFInputnodes could be GetShapesFromPlaceholder

Can be addressed in another PR

Reply:
Yes, this might change in later PRs. Right now it only accepts placeholders etc. But soon users might provide hints about "max" in the middle of the graph (say for the where op), so this section might need a redo

Reply:
ok, so node_names will be provided rather than type?

Reply:
the function `GetShapesFromTFInputnodes` takes a graph and a string. This string is basically "placeholder". So this function scans the graph for "placeholder" type nodes and finds what TF is saying about their shapes. Note `const string& input_node_type` as a separate input is probably not needed, Originally I had that as an argument as we get "placeholders" in case of grappler and "args" in case of normal pass.

AOT is supposed to work with grappler only, hence for now the only value input_node_type accepts is the string "placeholder". Maybe I can remove the argument. But its there for historical reasons, in case we want AOT to work with normal pass as well